core/example/solvers/pme.cpp

```diff
@@ -0,0 +1,449 @@
+}
+
+
+//Compute a 1D cubic cardinal B-spline value, used in spreading charge to mesh points
+//   Given the distance from the particle (x) in units of mesh spaces, this computes the 
+//   fraction of that charge to place at a mesh point x mesh spaces away
+//   The cubic B-spline used here is shifted so that it is symmetric about zero
+//   All cubic B-splines are smooth functions that go to zero and are defined piecewise
+KOKKOS_INLINE_FUNCTION
+double TPME::oneDspline(double x)
+{
+  if ( x >= 0.0 and x < 1.0 ) {
+     return (1.0/6.0) * x*x*x;
+  } 
+  else if ( x >= 1.0 and x <= 2.0 )
+  {
+     return -(1.0/2.0)*x*x*x + 2.0*x*x - 2.0*x + (2.0/3.0);
+  } 
+  //Using the symmetry here, only need to define function between 0 and 2
+  //Beware: This means all input to this function should be made positive
+  else
+  {
+     return 0.0;//Zero if distance is >= 2 mesh spacings
+  }
+} 
+
+//Compute a 1-D Euler spline. This function is part of the "lattice structure factor"
+//and is given by:
+//   b(k, meshwidth) = exp(2*PI*i*3*k/meshwidth) / SUM_{l=0,2}(1Dspline(l+1) * exp(2*PI*i*k*l/meshwidth))
+//   when using a non-shifted cubic B-spline in the charge spread, where meshwidth is the number of 
+//   mesh points in that dimension and k is the scaled fractional coordinate
+KOKKOS_INLINE_FUNCTION
+double TPME::oneDeuler(int k, int meshwidth)
+{ 
+  double denomreal = 0.0; 
+  double denomimag = 0.0;
+  //Compute the denominator sum first, splitting the complex exponential into sin and cos
+  for(int l = 0; l < 3; l++)
+  {
+     denomreal += TPME::oneDspline(std::min(4.0-(l+1.0),l+1.0)) * cos( 2.0 * PI * double(k) * l / double(meshwidth));
+     denomimag += TPME::oneDspline(std::min(4.0-(l+1.0),l+1.0)) * sin( 2.0 * PI * double(k) * l / double(meshwidth));
+  }
+  //Compute the numerator, again splitting the complex exponential
+  double numreal = cos(2.0*PI*3.0*double(k)/double(meshwidth));
+  double numimag = sin(2.0*PI*3.0*double(k)/double(meshwidth));
+  //Returning the norm of the 1-D Euler spline
+  return (numreal*numreal + numimag*numimag) / (denomreal*denomreal + denomimag*denomimag);
+}
+
+//Compute the energy
+void TPME::compute(ParticleList& particles, ParticleList& mesh, double lx, double ly, double lz)
+{
+  //Initialize energies: real-space, k-space (reciprocal space), self-energy correction, dipole correction
+  double Ur = 0.0, Uk = 0.0, Uself = 0.0, Udip = 0.0;
+  double Udip_vec[3];
+
+  //Particle slices
+  auto r = particles.slice<Position>();
+  auto q = particles.slice<Charge>();
+  auto p = particles.slice<Potential>();
+  
+  //Mesh slices
+  auto meshr = mesh.slice<Position>();
+  auto meshq = mesh.slice<Charge>();
+
+  //Number of particles
+  int n_max = particles.size();
+
+  total_energy = 0.0; 
+
+  //Set the potential of each particle to zero
+  auto init_p = KOKKOS_LAMBDA( const int idx )
+  {
+    p(idx) = 0.0;
```

Comment:
I've been thinking we a need a variant of `deep_copy` that assigns a scalar to every value in a slice. This is identical to the functionality available for `Kokkos::View`. This loop here would be a good candidate for that.

Reply:
I agree - deep_copy would be useful for the force reset in CabanaMD